### PR TITLE
Feature/65 screenshake

### LINKED ISSUE #65: Screenshake
## Description

The SHAKESCREEN action should be implemented. It will shake the screen with a given intensity for a given amount of time (note: this might take some creative interpretation and changing of the existing command).

## Acceptance Criteria

This story will be finished when the aforementioned action has been implemented and is testable in a unity scene

### PR DESCRIPTION
**Pull request by PhoebeMitchell**
## Summary

Adds "ScreenShaker" component which can be added to objects to shake them. ScreenShaker uses ScreenshakeCalculator class to get camera positions. ScreenshakeCalculator is tested with ScreenshakeCalculatorTests.

This PR sets the standard inclusion's UI Canvas to world space to allow shaking of the UI when the camera shakes.

## Before/after screenshots and/or animated gif

screenshake2

## Testing instructions

Play the ScreenshakeTest narrative script in Inky-TestScene. Run the PlayMode and EditMode tests.

## Additional information

- `[ ]` Changes UI
- `[x]` Introduces new feature
- `[ ]` Removes existing feature
- `[x]` Has associated resource: 
  
closes #65

### CODE REVIEW

File: Assets/Tests/PlayModeTests/Scripts/EvidenceMenu/EvidenceMenuMouseTests.cs [hunk 1]
```diff
@@ -13,9 +13,15 @@ public class EvidenceMenuMouseTests
...
-        
+
         private Keyboard Keyboard => _inputTestTools.Keyboard;
         private Mouse Mouse => _inputTestTools.Mouse;
+
+        [SetUp]
+        public void Setup()
+        {
+            
+        }
```
Comment (ViMaSter) [inline Assets/Tests/PlayModeTests/Scripts/EvidenceMenu/EvidenceMenuMouseTests.cs]: Nit: This method is empty and I think safe to removed.

File: Assets/Tests/PlayModeTests/Tools/InputTestTools.cs [hunk 2]
```diff
@@ -53,7 +53,7 @@ public IEnumerator PressForSeconds(ButtonControl control, float seconds, int rep
         /// <param name="position">The position to set the mouse to.</param>
         public IEnumerator SetMousePosition(Vector2 position)
         {
-            Set(Mouse.position, position);
+            Set(Mouse.position, Camera.main.WorldToScreenPoint(position));
```
Comment (ViMaSter) [inline Assets/Tests/PlayModeTests/Tools/InputTestTools.cs:107]: Keeping things reproducible and abstract. Good catch! Inside [`/Assets/Tests/PlayModeTests/Scripts/EvidenceMenu/EvidenceMenuMouseTests.cs`](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/pull/92/files#diff-5985ee5ed1e37579040cb5f2c6b22db59e31bdab860069cd864b11b510f48e93) too! 👍

File: Assets/Scripts/TextDecoder/DirectorActionDecoder.cs [hunk 3]
```diff
@@ -252,15 +252,35 @@ void ShakeScreen(string intensity)
             return;
 
         float intensityNumerical;
+        float duration;
```
Comment (ViMaSter) [inline Assets/Scripts/TextDecoder/DirectorActionDecoder.cs]: Nit: These two variables can be inlined:
1. Change `out intensityNumerical` (`line 266`) and `out duration` (`line 272`) to `out float intensityNumerical` and `out float duration` respectively
2. Remove line 254 and 255.

File: Assets/Scripts/TextDecoder/DirectorActionDecoder.cs [hunk 4]
```diff
@@ -252,15 +252,35 @@ void ShakeScreen(string intensity)
...
+
+        if (!float.TryParse(parameters[1], NumberStyles.Any, CultureInfo.InvariantCulture, out duration))
         {
-            Debug.LogError("Invalid paramater " + intensity + " for function SHAKESCREEN");
+            Debug.LogError("Invalid parameter " + parameters[1] + " for function SHAKESCREEN");
         }
+        
+        if (parameters.Length > 2 && !bool.TryParse(parameters[2], out shouldWaitForShake))
+        {
+            Debug.LogError("Invalid parameter " + parameters[2] + " for function SHAKESCREEN");
```
Comment (ViMaSter) [inline Assets/Scripts/TextDecoder/DirectorActionDecoder.cs]: Nit: I personally prefer using [string interpolation](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/interpolated) over concatenation when possible.

File: Assets/Scripts/Screenshake/ScreenshakeCalculator.cs [hunk 5]
```diff
@@ -0,0 +1,66 @@
+using System.Collections.Generic;
+using System.Linq;
```
Comment (ViMaSter) [inline Assets/Scripts/Screenshake/ScreenshakeCalculator.cs]: Nit: No symbols from these namespaces are used. I would recommend removing them from this file.

File: Assets/Scripts/Screenshake/ScreenshakeCalculator.cs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    private readonly Vector2 _noiseScale;
+    private readonly Vector2 _noiseOffset;
+    private readonly AnimationCurve _animationCurve;
+    private float _completion = 1;
+    private float _time;
+    
+    public bool IsShaking => !Mathf.Approximately(_completion, 0);
+
+    /// <summary>
+    /// Initialised private methods on object construction.
```
Comment (ViMaSter) [inline Assets/Scripts/Screenshake/ScreenshakeCalculator.cs]: Nit: This reads weird to me; did you mean `Initializes private members [...]`?

Generally I'd update this to reflect **why** the code is doing things over **what** is happening (I'd argue the code inside the constructor is pretty easy to read atm. and doesn't require documentation on **what** it's doing).  
A possible alternative could be: `Creates ScreenshakeCalculator instance with pre-defined intensity`

File: Assets/Scripts/Screenshake/ScreenShaker.cs [hunk 7]
```diff
@@ -0,0 +1,79 @@
+using System.Collections;
+using UnityEngine;
+using UnityEngine.Events;
+using UnityEngine.Serialization;
```
Comment (ViMaSter) [inline Assets/Scripts/Screenshake/ScreenShaker.cs]: Nit: No symbols from this namespace are used. I would recommend removing it from this file.

File: Assets/Scripts/Screenshake/ScreenShaker.cs [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    {
+        _transform = transform;
+        _cameraOffset = _transform.position;
+    }
+
+    /// <summary>
+    /// Call this method to begin shaking screen.
+    /// </summary>
+    /// <param name="intensity">The time (in seconds) that the shake will last.</param>
+    /// <param name="shouldWaitForShake">Whether the system waits for the shake to complete before continuing.</param>
```
Comment (ViMaSter) [inline Assets/Scripts/Screenshake/ScreenShaker.cs:40]: Nit: `<param>`-comment for `time` parameter is missing.

**Review by ViMaSter — DISMISSED** [review]
I've definitely seen a few files like `Assets/Scripts/Screenshake/ScreenShaker.cs` in game code.  
And I've also seen a lot of files like `Assets/Tests/EditModeTests/ScreenshakeCalculatorTests.cs` in a few enterprise-level codebases.  
But I don't think I've ever seen both types of code, in a game project and coming from the same person.

This is absolutely rock solid logic, great test coverage, a super clean and simple interface to use and an insanely important feature to make our game "feel right"! Crazy well done man! 👏 

---

(Still doesn't stop me from finding nits to pick though! 😛)

File: Assets/Scripts/BGScenes/SceneController.cs [hunk 9]
```diff
@@ -149,9 +152,20 @@ public void SetCameraPos(Vector2Int position)
...
     }
 
-    public void ShakeScreen(float intensity)
+    /// <summary>
+    /// Calls shake method on the assigned screen shaker.
+    /// </summary>
+    /// <param name="intensity">The intensity of the shake.</param>
+    /// <param name="duration">The duration of the shake.</param>
+    /// <param name="shouldWaitForShake">Whether the system waits for the shake to complete before continuing.</param>
+    public void ShakeScreen(float intensity, float duration, bool shouldWaitForShake)
```
Comment (notexplosive) [inline Assets/Scripts/BGScenes/SceneController.cs]: nit: `shouldWaitForShake` should be named something like `isBlocking`

**Review by notexplosive — DISMISSED** [review]
Lookin good!

File: Assets/Tests/PlayModeTests/Scripts/EvidenceMenu/EvidenceMenuMouseTests.cs (hunk 1 above)
Comment (PhoebeMitchell) [reply Assets/Tests/PlayModeTests/Scripts/EvidenceMenu/EvidenceMenuMouseTests.cs]: Done

File: Assets/Scripts/TextDecoder/DirectorActionDecoder.cs (hunk 3 above)
Comment (PhoebeMitchell) [reply Assets/Scripts/TextDecoder/DirectorActionDecoder.cs]: Done

File: Assets/Scripts/TextDecoder/DirectorActionDecoder.cs (hunk 4 above)
Comment (PhoebeMitchell) [reply Assets/Scripts/TextDecoder/DirectorActionDecoder.cs]: Good point. I just copy pasted the existing error message. Changed all of them in this method to use interpolation.

File: Assets/Scripts/Screenshake/ScreenShaker.cs (hunk 7 above)
Comment (PhoebeMitchell) [reply Assets/Scripts/Screenshake/ScreenShaker.cs]: Done

File: Assets/Scripts/Screenshake/ScreenShaker.cs (hunk 8 above)
Comment (PhoebeMitchell) [reply Assets/Scripts/Screenshake/ScreenShaker.cs:40]: Added, and changed `time` to `duration` to be consistent with other method parameters. Also updated `intensity` description.

File: Assets/Scripts/Screenshake/ScreenshakeCalculator.cs (hunk 5 above)
Comment (PhoebeMitchell) [reply Assets/Scripts/Screenshake/ScreenshakeCalculator.cs]: Done

File: Assets/Scripts/Screenshake/ScreenshakeCalculator.cs (hunk 6 above)
Comment (PhoebeMitchell) [reply Assets/Scripts/Screenshake/ScreenshakeCalculator.cs]: Yep, sleep deprived moment 😄 Amy likes to have summaries so changed to your suggestion.

File: Assets/Scripts/BGScenes/SceneController.cs (hunk 9 above)
Comment (PhoebeMitchell) [reply Assets/Scripts/BGScenes/SceneController.cs]: Changed to isBlocking.

**Review by lyndsiWilliams — APPROVED** [review]
Code is looking shiny! Tested this locally and it works and looks great!

**Review by ViMaSter — APPROVED** [review]
Rock solid, looks good to me now!

Was happy to see that nits from other PRs were also taken into consideration; [the parameter name consistency](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/pull/92/commits/04da625900d9d4a3db9252b2dba396d4ecff7c8e#diff-24f1b69f75ec838fa74f3c0f78b671b8df0cd85149610bea2b062b41232ddb54R158-R165) for everything related to `ShakeScreen` you've introduced in 04da6259 is great! 🚀

**Pull request merged**